aws-s3-transfer-manager/src/types.rs

```diff
@@ -179,7 +179,20 @@ fn is_hidden(path: &Path) -> bool {
 impl Default for UploadFilter {
     fn default() -> Self {
         Self {
-            predicate: Arc::new(|item| item.metadata().is_file() && !is_hidden(item.path())),
+            predicate: Arc::new(|item| {
+                if is_hidden(item.path()) {
+                    return false;
+                }
+                if item.symlink_metadata.is_symlink() {
```

Comment:
If our transfer manager's `follow_symlinks` setting is `false`, should it stop us from uploading symlinked files? Or is only supposed to stop us from recursing *into* symbolic directories?

If it's supposed to stop us from uploading symbolic files, then we shouldn't need to have this check here in the `UploadFilter`. Our code should have rejected this DirEntry before presenting it to the user's `UploadFilter`

I tested `WalkDir` and its [WalkDir.follow_links(false)](https://docs.rs/walkdir/latest/walkdir/struct.WalkDir.html#method.follow_links) only stops it from *entering inside* symbolic directories. It will still yield the symbolic directories and files as it iterates.

Reply:
> should it stop us from uploading symlinked files? Or is only supposed to stop us from recursing into symbolic directories?

Good question. The SEP doesn't seem to specify this behavior, but according to a [Java SDK V2's test](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L248-L272), it's the former (symlinked files are [not in the expected test result](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L271)) and so is the implementation in this PR.

> If it's supposed to stop us from uploading symbolic files, then we shouldn't need to have this check here in the UploadFilter

We actually need this check, because a symlink can be a link to a single file, in which case, a `DirEntry` representing the symlink to the single file is still presented to `UploadFilter`. In fact, if we removed `if item.symlink_metadata.is_symlink()`, I believe [this unit test](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/08330ab0bb1fb4d9a4c62595d2525d2f0e2ca8e9/aws-s3-transfer-manager/tests/upload_objects_test.rs#L204-L218) would fail because `output.objects_uploaded()` would become 2, including `symlink2` as a file to upload as well.

Reply:
> should it stop us from uploading symlinked files? Or is only supposed to stop us from recursing into symbolic directories?

Good question. The SEP doesn't seem to specify this behavior, but according to a [Java SDK V2's test](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L248-L272), it's the former (symlinked files are [not in the expected test result](https://github.com/aws/aws-sdk-java-v2/blob/master/services-custom/s3-transfer-manager/src/test/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryHelperTest.java#L271)) and so is the implementation in this PR.

> If it's supposed to stop us from uploading symbolic files, then we shouldn't need to have this check here in the UploadFilter

I'm pretty sure we need this check, because a symlink can be a link to a single file, in which case, a `DirEntry` representing the symlink to the single file is still presented to `UploadFilter`. If we removed `if item.symlink_metadata.is_symlink()` and only left `item.symlink_metadata.is_file()` as the last expression, I believe [this unit test](https://github.com/awslabs/aws-s3-transfer-manager-rs/blob/08330ab0bb1fb4d9a4c62595d2525d2f0e2ca8e9/aws-s3-transfer-manager/tests/upload_objects_test.rs#L186-L202) would fail because `output.objects_uploaded()` would become 6, incorrectly rejecting `symlink2` as `item.symlink_metadata.is_file()` returns false (and `item.symlink_metadata.is_symlink()` returns true).

Reply:
I get that the `is_symlink()` check needs to happen. But I'm saying the Transfer Manager should be checking this *before* presenting the item to the user-written `UploadFilter`.

Since `WalkDir` is going to present the Transfer manager with symlinked files even if when `follow_symlinks` is false, it's up to the Transfer Manager to skip over these without relying on the user's `UploadFilter`.

Reply:
And if we're not relying on the `UploadFilter` to deal with symlinks, then we could simplify `UploadFilterItem` to only have the actual `metadata` (vs having both `symlink_metadata` and optional `metadata`)

Reply:
Ah, got it. Thanks for clarifying. Yes, that should help simplify things. I'll make that adjustment.